File: Habibi/src/api/types/inbox.ts
/**
 * Domain / wire types for the inbox surface.
 *
 * The wire shape the api/ module for this surface reads and writes.
 */

// Mirrors the conversations.channel CHECK constraint (sql/04_interactions.sql).
// Narrower than the database, the inbox drops threads it was meant to show.
export type Channel = "whatsapp" | "sms" | "email" | "chat" | "voice";
export type Sender = "customer" | "bot" | "agent" | "system";
/** Stored conversation status — "mine" is derived (assignedUserId === me). */
export type ThreadStatus = "bot" | "needs_human" | "escalated" | "assigned";
export type SlaLevel = "ok" | "warn" | "breach";
export type Sentiment = "positive" | "neutral" | "negative";
// "pending" = queued by the API, not yet accepted by the provider. It must be
// distinguishable from a delivered message: rendering both as "no tick" is how
// an agent spent six minutes replying to a customer who saw nothing.
export type DeliveryStatus = "pending" | "sent" | "delivered" | "read" | "failed";
export interface Message {
  id: string;
  sender: Sender;
  text: string;
  time: string; // "3:41 PM"
  delivery?: DeliveryStatus; // for bot/agent
}
export interface SystemEvent {
  id: string;
  kind: "system";
  text: string;
  time: string;
}
export type ThreadItem = Message | SystemEvent;
export interface ThreadContext {
  riskLevel: "High" | "Medium" | "Low";
  contactableNow: boolean;
  contactWindow: string;
  outstanding: number;
  outstandingAging: string;
  nextEmiDate: string;
  nextEmiAmount: number;
  lastPromise: {
    amount: number;
    date: string;
    status: "Kept" | "Broken" | "Pending" | "Partial";
  } | null;
  openDisputes: { id: string; summary: string }[];
  recentInteractions: {
    id: string;
    kind: "call" | "chat";
    summary: string;
    when: string;
    sentiment: Sentiment;
  }[];
}
export interface Thread {
  id: string;
  customer: string;
  /** CRM customer id for Customer 360 / PTP / dispute deep-links. */
  customerId?: string;
  accountId: string;
  channel: Channel;
  status: ThreadStatus;
  /** Owning agent; Mine filter = assignedUserId === current user. */
  assignedUserId: string | null;
  isMine: boolean;
  /** True while a bot turn is queued/running (WhatsApp auto-reply in flight). */
  botTyping?: boolean;
  /** True while bot/agent outbound WhatsApp send is queued or mid-flight. */
  pendingOutbound?: boolean;
  /** Server watermark for delta polls (`?updatedAfter=`). */
  updatedAt?: string | null;
  sla: SlaLevel;
  unread: number;
  lastTime: string;
  lastPreview: string;
  lastFrom: Sender;
  sentiment: Sentiment;
  handlerBotId?: string | null;
  ragSuggestions: string[];
  /** Optional grounded draft from shared kb_retrieve (same as Test Retrieval). */
  ragDraftAnswer?: string | null;
  messages: ThreadItem[];
  /** Present on GET /conversations/{id} and write responses; the list omits it. */
  context?: ThreadContext | null;
}

File: backend/schemas/inbox.py
"""The conversation inbox.

One module of the ``schemas`` package (was one 6,400-line file); the
router of the same name serves these. ``schemas/__init__`` re-exports every name.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# The authored flow graph is a domain model, not a transport shape — it is
# shared verbatim by the API, the validator and the voice runtime, so it is
# defined once in flow_graph and reused here rather than restated.
from flow_graph import FlowGraph, FlowIssue, FlowValidation  # noqa: F401

class HandoffDisclosureRequest(BaseModel):
    itemId: str
    ruleId: str | None = None
    label: str | None = None
    read: bool = True


# ---------------------------------------------------------------------------
# Conversation Inbox (Phase 3B Tier 3)
# ---------------------------------------------------------------------------


class InboxMessageResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    sender: Literal["customer", "bot", "agent"]
    text: str
    time: str
    # "pending" = accepted by the API and queued, not yet handed to the
    # provider. Its own state on purpose — see db._inbox_delivery.
    delivery: Literal["pending", "sent", "delivered", "read", "failed"] | None = None


class InboxSystemEventResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    kind: Literal["system"] = "system"
    text: str
    time: str


class InboxPromiseResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    amount: float
    date: str
    status: Literal["Kept", "Broken", "Pending", "Partial"]


class InboxDisputeSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    summary: str


class InboxInteractionSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    kind: Literal["call", "chat"]
    summary: str
    when: str
    sentiment: Literal["positive", "neutral", "negative"]


class InboxThreadContextResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    riskLevel: Literal["High", "Medium", "Low"]  # `critical` on the book reads High here
    contactableNow: bool
    contactWindow: str
    outstanding: float
    outstandingAging: str
    nextEmiDate: str
    nextEmiAmount: float
    lastPromise: InboxPromiseResponse | None = None
    openDisputes: list[InboxDisputeSummaryResponse] = []
    recentInteractions: list[InboxInteractionSummaryResponse] = []


class ConversationListResponse(BaseModel):
    """Conversation Inbox screen Thread shape."""

    model_config = ConfigDict(extra="forbid")

    id: str
    customer: str
    customerId: str
    accountId: str
    # Every value the conversations.channel CHECK constraint permits, and the
    # constraint is the authority — see tests/test_inbox_channel_contract.py.
    #
    # This listed three of the five. `response_model` validates the WHOLE list,
    # so the first voice conversation ever written did not render as an odd row:
    # it raised ResponseValidationError and took the entire inbox down with
    # "Failed to load inbox: Failed to fetch". One sandbox call was enough.
    channel: Literal["whatsapp", "sms", "email", "chat", "voice"]
    status: Literal["bot", "needs_human", "escalated", "assigned"]
    assignedUserId: str | None = None
    isMine: bool
    botTyping: bool = False
    pendingOutbound: bool = False
    updatedAt: str | None = None
    sla: Literal["ok", "warn", "breach"]
    unread: int
    lastTime: str
    lastPreview: str
    lastFrom: Literal["customer", "bot", "agent"]
    sentiment: Literal["positive", "neutral", "negative"]
    ragSuggestions: list[str] = []
    ragDraftAnswer: str | None = None
    handlerBotId: str | None = None
    messages: list[InboxMessageResponse | InboxSystemEventResponse] = []
    context: InboxThreadContextResponse


class CannedResponseItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    label: str
    text: str


class ConversationMessageCreateRequest(BaseModel):
    text: str = Field(min_length=1)


class ConversationSuggestionsRefreshRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    topK: int = Field(default=4, ge=1, le=20)
    includeDraftAnswer: bool = False


class ConversationSuggestionsRefreshResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    conversationId: str
    ragSuggestions: list[str]
    draftAnswer: str | None = None
    chatModel: str | None = None
    latencyMs: int | None = None
    logId: str | None = None
    thread: ConversationListResponse | None = None
